README.txt
************************************
*  JFREECHART: Version 1.2.0-pre1  *
************************************

19 December 2008

(C)opyright 2000-2008, by Object Refinery Limited and Contributors.

-----------------
1.  INTRODUCTION
-----------------
JFreeChart is a free chart library for the Java(tm) platform.  It runs 
on the Java 2 Platform (JDK 1.4.2 or later) and uses the Java 2D API for 
drawing.

JFreeChart is licensed under the terms of the GNU Lesser General
Public Licence (LGPL).  A copy of the licence is included in the
distribution.

Please note that JFreeChart is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  Please refer to the licence for details.

-------------------
2.  LATEST VERSION
-------------------
The latest version of this class library can be obtained from:

    http://www.jfree.org/jfreechart/

If you have an comments, suggestions or bugs to report, please post a
message in the JFreeChart forum.

-----------------
3.  DOCUMENTATION
-----------------
You can download installation instructions
(jfreechart-1.2.0-install.pdf) from the JFreeChart home
page or the project page on SourceForge.

Further documentation for JFreeChart (the JFreeChart Developer Guide) is
available to purchase from by Object Refinery Limited, a company owned and
operated by David Gilbert (the original author of JFreeChart).  For more
information, please see:

    http://www.object-refinery.com/jfreechart/guide.html

Sales of the JFreeChart Developer Guide are an important source of funding for
the project.  Please help us to continue developing free software.

API documentation files in HTML format are available on-line.  If you wish 
to keep a local copy of the API files, you can regenerate them directly
from the source code.  You can do this using the 'javadoc' or 'gjdoc' 
utilities directly, or with the Ant script (build.xml) included in the
distribution.

-----------------
4.  DEPENDENCIES
-----------------
JFreeChart has the following dependencies:

(a)  JDK 1.4.2 or higher - if you are using JFreeChart to create applets, 
this means that you cannot rely on the JVM integrated with Microsoft's 
Internet Explorer - your users will need to have the Java 2 plug-in 
installed.  Most other browsers (e.g. Firefox, Mozilla, Netscape, Konqueror) 
support JRE 1.4, 1.5 or 1.6.

(b)  servlet.jar - classes in the org.jfree.chart.servlet
package require this file.  The JFreeChart distribution includes the 
servlet.jar file distributed with Tomcat 4.1.31.  Applicable license 
terms are published at:  

    http://java.sun.com/products/servlet/LICENSE   

(c)  JUnit - a unit testing framework (the junit.jar runtime file is
included in the distribution).  JUnit is licensed under the terms
of the IBM Common Public License.  You can find out more about JUnit
and/or download the latest version from:

    http://www.junit.org

The JUnit tests included with JFreeChart have been created using JUnit
4.3.1.

-----------
5.  SUPPORT
-----------
Support questions can be posted in the free support forum at

    http://www.jfree.org/phpBB2/viewforum.php?f=3
    
We read all questions posted in the forum, and respond to as many as we can
in the time available.  Unfortunately, there are too many questions to answer
them all.
    
If you need commercial support for JFreeChart, this is offered by
Object Refinery Limited:

    http://www.object-refinery.com/jfreechart/support.html

--------------------
6.  ANT BUILD SCRIPT
--------------------
An Ant build script (build.xml) is included in the distribution.  This
is the same script that is used to create the JFreeChart distribution.

For more information about Ant:

    http://ant.apache.org/

Please note that you will need to obtain the servlet.jar file (see the
DEPENDENCIES section above) before running the Ant script.

------------------------
7.  THE DEMO APPLICATION
------------------------
A demo application that shows a selection of the charts that can be
generated is included in the JFreeChart distribution.   To run the
demo (using JDK 1.4.2 or later), use the following command:

    java -jar jfreechart-1.2.0-pre1-demo.jar

The complete source code for the demo application is available for
download when you purchase the JFreeChart Developer Guide.

The demo application uses iText, a library for creating PDF documents.
iText is licensed under the terms of the GNU Lesser General Public Licence
and is available to download from:

    http://www.lowagie.com/iText/

---------------
8.  LIMITATIONS
---------------
JFreeChart has some known limitations that will hopefully be addressed in 
the future:

    - with the XYPlot class, when the bounds of the range axis are calculated
      automatically, all of the data is used, not just the subset that is 
      visible on the chart;
    - some renderers do not respect the series visibility flags yet;
    - the chart property editors (accessible by right-clicking on the chart
      panel) are horribly out of date and probably shouldn't be used;
    - item labels (if displayed) are not taken into account for the 
      automatically calculated axis range.  As a workaround, you can increase
      the axis margins;
    - tick labels on a DateAxis that uses a SegmentedTimeline can be 
      problematic;
    
If there are other items that you think should be listed here,
please post a bug report.

-----------------
9.  GNU CLASSPATH
-----------------
GNU Classpath is a project to create a free, clean-room implementation of the
standard Java(tm) class libraries:

http://www.gnu.org/software/classpath/

JFreeChart should work quite well with free runtimes based on 
GNU Classpath 0.92 or later.  We are grateful to all the GNU Classpath 
developers that have made this possible.

---------------
10.  WHAT's NEW
---------------
A list of changes in recent versions:

1.2.0-pre1 : (25-Jun-2007)
        - removed JCommon dependency;
        - removed support for JDK 1.3.1;
        - removed deprecated code.

1.0.6 : (15-Jun-2007) 
        - the VectorRenderer and associated dataset classes have been promoted
          to the standard API from the 'experimental' source tree.  See the
          NEWS and ChangeLog files for a more detailed description of the 
          changes in this release.
          
1.0.5 : (23-Mar-2007) 
        - this release contains a new DeviationRenderer, enhancements to a 
          number of existing classes and numerous bug fixes, see the NEWS 
          and ChangeLog files for details.

1.0.4 : (9-Feb-2007)
        - this release contains both new features and bug fixes, see the NEWS 
          and ChangeLog files for details.

1.0.3 : (17-Nov-2006) 
        - this release contains a new DialPlot class (in 'experimental') that
          will hopefully replace the MeterPlot class, once the API has been
          polished a little, plus many other new features and bug fixes.  See
          the NEWS and ChangeLog files for details.

1.0.2 : (25-Aug-2006)
        - this release contains both new features and bug fixes, see the NEWS 
          and ChangeLog files for details.

1.0.1 : (27-Jan-2006)
        - primarily a bug fix release, see the NEWS and ChangeLog files for
          details.

1.0.0 : (2-Dec-2005)
        - the first stable release of the JFreeChart class library, all future
          releases in the 1.0.x series will aim to maintain backward 
          compatibility with this release;
        - see the ChangeLog file for a detailed list of changes.

1.0.0-rc3 : (28-Nov-2005)
        - the third "release candidate" for version 1.0.0, this release
          fixes some issues with the 1.0.0-rc2 release (mainly concerning
          packaging of resource bundles for localisation).  
        - if no significant problems are reported in the next few days,
          the 1.0.0 "final" release will be posted on 2-Dec-2005.

1.0.0-rc2 : (25-Nov-2005)
            - the second "release candidate" for version 1.0.0.  If no problems
              are reported, 1.0.0 "final" will be released on 2-Dec-2005.
            - see the ChangeLog file for a detailed list of changes in this 
              release.

1.0.0-rc1 : (2-Jun-2005)
            - this is a "release candidate" for version 1.0.0.  If no 
              significant API problems are reported, this release will be 
              re-released as version 1.0.0. 
            - see CHANGELOG.txt for further details.

1.0.0-pre2 : (10-Mar-2005)
         - see CHANGELOG.txt for further details.

1.0.0-pre1 : (29-Nov-2004)
         - see CHANGELOG.txt for further details.
 
0.9.21 : (9-Sep-2004) 
         - added new axes: PeriodAxis and ModuloAxis.
         - split org.jfree.data and org.jfree.chart.renderer into
           subpackages for 'category' and 'xy' charts.
         - Sun PNG encoder is now used, if available.
         - a new demo application makes it easier to preview the 
           chart types that JFreeChart can create.
         - added a new series visibility flag to the AbstractRenderer
           class.
         - added support for GradientPaint in interval markers.
         - see CHANGELOG.txt for further details.
         
0.9.20 : (7-Jun-2004) 
         - primarily bug fixes, see CHANGELOG.txt for further details.

0.9.19 : (28-May-2004)
         - added methods to XYDataset that return double primitives;
         - removed distinction between "primary" and "secondary" datasets,
           renderers and axes;
         - added fixed legend item options to CategoryPlot and XYPlot;
         - legend changes by Barek Naveh;
         - removed Log4j dependency;
         - many, many bug fixes;
         - see CHANGELOG.txt for further details.

0.9.18 : (15-Apr-2004)
         - new legend anchor options;
         - fixed broken JPEG export;
         - fixed title size problems;
         - various other bug fixes;

0.9.17 : (26-Mar-2004)
         - pie chart enhancements for labelling, shading and multiple pie
           charts (2D or 3D) on a single plot;
         - new PolarPlot class added;
         - XYSeries can now be sorted or unsorted;
         - createBufferedImage() method can now scale charts;
         - domain and range markers now support intervals;
         - item labels are now supported by some XYItemRenderers;
         - tooltip and item label generators now use MessageFormat class;
         - added new XYBarDataset class;
         - added transparency support to PNG export;
         - numerous other small enhancements and bug fixes, see the
           CHANGELOG.txt file for more details;

0.9.16 : (09-Jan-2004)
         - this release contains bug fixes and some minor feature
           enhancements (title and category label wrapping, legend shape
           scaling, enhanced performance for the DefaultTableXYDataset class);
         - added Spanish localisation files;

0.9.15 : (28-Nov-2003)
         - the focus of this release is bug fixes - quite a number of issues
           have been resolved, please check the bug database for details;
         - added a new Wafer Map chart type;
         - added a cyclic axis;
         - added localisation files for _ru;

0.9.14 : (17-Nov-2003) 
         - implemented zooming for the FastScatterPlot class;
         - added item label support for stacked bar charts, and new fall back
           options for item labels that don't fit within bars;
         - modified the CategoryAxis class to allow additional options for the
           alignment and rotation of category labels;
         - addition of the AxisState class, used in the drawing of axes to
           eliminate a bug when multiple threads draw the same axis
           simultaneously;
         - provided additional attributes in the DateTickUnit class to improve
           labelling on a segmented DateAxis;
         - added support for GradientPaint in bar charts;
         - updated the PNGEncoder;
         - fixes for tick label positioning on axes;
         - various Javadoc updates;
         - numerous bug fixes;

0.9.13 : (26-Sep-2003)
         - various enhancements to the stacked area XY charts;
         - added a completion indicator for the Gantt chart;
         - range and domain markers can now be placed in the foreground or the
           background;
         - more fixes for cloning and serialization;
         - fixed mouse event bug for combined charts;
         - fixed bugs in the PngEncoder class;
         - incorporated .properties files that were missing from the 0.9.12
           distribution;

0.9.12 : (11-Sep-2003)
         - extended box-and-whisker plots to work with the CategoryPlot	class
           as well as the XYPlot class (based on work by David Browning);
         - added a new LayeredBarRenderer (by Arnaud Lelievre);
         - added support for stacked area charts with the XYPlot class (thanks
           to Richard Atkinson);
         - improved HTML image map support (thanks to Richard Atkinson);
         - added localized resources for the chart property editors (thanks to
           Arnaud Lelievre).  Current translations include French and Portugese

ChangeLog
2006-06-16  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/chart/renderer/xy/YIntervalRenderer.java
    (additionalItemLabelGenerator): New field,
    (YIntervalRenderer): Initialise new field,
    (getAdditionalItemLabelGenerator): New method,
    (setAdditionalItemLabelGenerator): Likewise,
    (drawItem): Added additional item label support,
    (drawAdditionalItemLabel): New method,
    (equals): Include new field.

2006-06-16  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/chart/renderer/xy/HighLowRenderer.java
    (tickLength): New field,
    (HighLowRenderer): Initialise tickLength,
    (getTickLength): New method,
    (setTickLength): Likewise,
    (findRangeBounds): New method override,
    (drawItem): Use new tickLength field,
    (equals): Check new field.

2006-06-16  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/chart/labels/IntervalXYItemLabelGenerator.java: 
    New class.

2006-06-16  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/chart/labels/AbstractXYItemLabelGenerator.java
    (nullXString): Removed,
    (getNullYString): New method,
    (createItemArray): Removed redundant code,
    (equals): Check nullYString field,
    (hashCode): Implemented.

2006-06-16  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/chart/renderer/xy/XYSplineAndShapeRenderer.java:
    Renamed XYSplineRenderer.java.

2008-06-10  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/data/time/TimeSeriesCollection.java
    (getSeries): Changed argument from String to Comparable.

2008-06-10  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/data/time/Millisecond.java
    (compareTo): Added handling for general RegularTimePeriod.
    
2008-06-10  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/data/time/TimePeriodValue.java
    (toString): New method override,
    * source/org/jfree/data/time/TimePeriodValues.java
    (updateBounds): Update maxMiddleIndex correctly.
    
2008-06-10  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/data/time/Week.java
    (Week(Date, TimeZone)): Hand on the zone.
    
2008-06-10  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/data/time/SimpleTimePeriod.java
    (start): Changed from Date to long,
    (end): Likewise,
    (getStartMillis): New method,
    (getEndMillis): Likewise,
    (getStart): Returns new date instance,
    (getEnd): Likewise.

2008-06-10  David Gilbert  <david.gilbert@object-refinery.com>

    Bug 1864222:
    * source/org/jfree/data/time/TimeSeries.java
    (createCopy(RegularTimePeriod, RegularTimePeriod)): Handle empty range.
    
2008-06-10  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/data/time/FixedMillisecond.java
    (time): Changed from Date to long.

2008-06-10  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/data/time/DateRange.java
    (lowerDate): Changed from Date to long,
    (upperDate): Likewise,
    (DateRange(Date, Date)): Updated for field type change,
    (DateRange(long, long)): Likewise,
    (getLowerDate): Likewise,
    (getUpperDate): Likewise,
    (toString): Likewise.

2008-06-10  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/chart/axis/DateTickUnit.java
    (addToDate(Date)): Deprecated.

2008-06-10  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/chart/imagemap/DynamicDriveToolTipTagFragmentGenerator.java
    (generateToolTipFragment): Escape text,
    * source/org/jfree/chart/imagemap/OverLIBToolTipTagFragmentGenerator.java
    (generateToolTipFragment): Likewise,
    * source/org/jfree/chart/imagemap/StandardToolTipTagFragmentGenerator.java
    (generateToolTipFragment): Likewise.

2008-06-02  David Gilbert  <david.gilbert@object-refinery.com>

    * source/org/jfree/chart/renderer/xy/AbstractXYItemRenderer.java
    (isPointInRect): New utility method,
    * source/org/jfree/chart/renderer/xy/StandardXYItemRenderer.java
    (drawItem): Modified bounds check for inclusion of chart entity,
    * source/org/jfree/chart/renderer/xy/XYLineAndShapeRenderer.java
    (drawSecondaryPass): Modified bounds check for inclusion of chart entity.

2008-06-02  David Gilbert  <david.gilbert@object-refinery.com>

    * swt/org/jfree/experimental/chart/swt/ChartComposite.java
    (getChartRenderingInfo): New method,
    (isDomainZoomable): Likewise,
    (setDomainZoomable): Likewise,
    (isRangeZoomable): Likewise,
    (setRangeZoomable): Likewise.

2008-06-02  David Gilbert  <david.gilbert@object-refinery.com>

    Bug 1977609:
    * source/org/jfree/chart/util/SerialUtilities.java
    (readShape): Handle SEG_CLOSE for GeneralPath,
    * source/org/jfree/chart/util/ShapeUtilities.java
    (equal(GeneralPath, GeneralPath)): Use correct path for comparison,
    * tests/org/jfree/chart/util/junit/SerialUtilities.java: New file,
    * tests/org/jfree/chart/util/junit/ShapeUtilities.java: Likewise,
    * tests/org/jfree/chart/util/junit/UtilPackageTests.java: Include new tests.

2008-05-30  Henry Proudhon  <henry.proudhon@ensmp.fr>

    * source/swt/org/jfree/experimental/chart/swt/ChartComposite.java: 
    (ChartComposite): removed duplicate code.

2008-05-24  Henry Proudhon  <henry.proudhon@ensmp.fr>

    * source/swt/org/jfree/experimental/chart/swt/ChartComposite.java: 
    (paintControl): Corrected the axis traces which were inverted,
    (chart): Fixed listener registered twice,
    (dispose): release the buffer image resources.
    * source/swt/org/jfree/experimental/swt/SWTGraphics2D.java
    (setTransform): use resource from pool if possible,
    (getSwtTransformFromPool): New method,
    (disposeResourcePool): dispose transform resources.

2008-04-18  David Gilbert  <david.gilbert@object-refinery.com>

    Patch 1943021 (Brian Cabana):

source/org/jfree/data/time/Month.java
/* ===========================================================
 * JFreeChart : a free chart library for the Java(tm) platform
 * ===========================================================
 *
 * (C) Copyright 2000-2007, by Object Refinery Limited and Contributors.
 *
 * Project Info:  http://www.jfree.org/jfreechart/index.html
 *
 * This library is free software; you can redistribute it and/or modify it 
 * under the terms of the GNU Lesser General Public License as published by 
 * the Free Software Foundation; either version 2.1 of the License, or 
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, 
 * USA.  
 *
 * [Java is a trademark or registered trademark of Sun Microsystems, Inc. 
 * in the United States and other countries.]
 *
 * ----------
 * Month.java
 * ----------
 * (C) Copyright 2001-2007, by Object Refinery Limited and Contributors.
 *
 * Original Author:  David Gilbert (for Object Refinery Limited);
 * Contributor(s):   Chris Boek;
 *
 * $Id: Month.java,v 1.7.2.6 2007/04/04 10:58:19 mungady Exp $
 *
 * Changes
 * -------
 * 11-Oct-2001 : Version 1 (DG);
 * 14-Nov-2001 : Added method to get year as primitive (DG);
 *               Override for toString() method (DG);
 * 18-Dec-2001 : Changed order of parameters in constructor (DG);
 * 19-Dec-2001 : Added a new constructor as suggested by Paul English (DG);
 * 29-Jan-2002 : Worked on the parseMonth() method (DG);
 * 14-Feb-2002 : Fixed bugs in the Month constructors (DG);
 * 26-Feb-2002 : Changed getStart(), getMiddle() and getEnd() methods to 
 *               evaluate with reference to a particular time zone (DG);
 * 19-Mar-2002 : Changed API for TimePeriod classes (DG);
 * 10-Sep-2002 : Added getSerialIndex() method (DG);
 * 04-Oct-2002 : Fixed errors reported by Checkstyle (DG);
 * 10-Jan-2003 : Changed base class and method names (DG);
 * 13-Mar-2003 : Moved to com.jrefinery.data.time package, and implemented 
 *               Serializable (DG);
 * 21-Oct-2003 : Added hashCode() method (DG);
 * 01-Nov-2005 : Fixed bug 1345383 (argument check in constructor) (DG);
 * ------------- JFREECHART 1.0.x ---------------------------------------------
 * 05-Oct-2006 : Updated API docs (DG);
 * 06-Oct-2006 : Refactored to cache first and last millisecond values (DG);
 * 04-Apr-2007 : Fixed bug in Month(Date, TimeZone) constructor (CB);
 * 21-Jun-2007 : Removed JCommon dependencies (DG);
 *
 */

package org.jfree.data.time;

import java.io.Serializable;
import java.util.Calendar;
import java.util.Date;
import java.util.TimeZone;

import org.jfree.date.SerialDate;

/**
 * Represents a single month.  This class is immutable, which is a requirement
 * for all {@link RegularTimePeriod} subclasses.
 */
public class Month extends RegularTimePeriod implements Serializable {

    /** For serialization. */
    private static final long serialVersionUID = -5090216912548722570L;

    /** The month (1-12). */
    private int month;

    /** The year in which the month falls. */
    private int year;

    /** The first millisecond. */
    private long firstMillisecond;
    
    /** The last millisecond. */
    private long lastMillisecond;

    /**
     * Constructs a new Month, based on the current system time.
     */
    public Month() {
        this(new Date());
    }

    /**
     * Constructs a new month instance.
     *
     * @param month  the month (in the range 1 to 12).
     * @param year  the year.
     */
    public Month(int month, int year) {
        if ((month < 1) || (month > 12)) {
            throw new IllegalArgumentException("Month outside valid range.");
        }
        this.month = month;
        this.year = year;
        peg(Calendar.getInstance());
    }

    /**
     * Constructs a new month instance.
     *
     * @param month  the month (in the range 1 to 12).
     * @param year  the year.
     */
    public Month(int month, Year year) {
        if ((month < 1) || (month > 12)) {
            throw new IllegalArgumentException("Month outside valid range.");
        }
        this.month = month;
        this.year = year.getYear();
        peg(Calendar.getInstance());
    }

    /**
     * Constructs a new <code>Month</code> instance, based on a date/time and 
     * the default time zone.
     *
     * @param time  the date/time.
     */
    public Month(Date time) {
        this(time, RegularTimePeriod.DEFAULT_TIME_ZONE);
    }

    /**
     * Constructs a new <code>Month</code> instance, based on a date/time and 
     * a time zone.  The first and last millisecond values are initially
     * pegged to the given time zone also.
     *
     * @param time  the date/time.
     * @param zone  the time zone (<code>null</code> not permitted).
     */
    public Month(Date time, TimeZone zone) {
        Calendar calendar = Calendar.getInstance(zone);
        calendar.setTime(time);
        this.month = calendar.get(Calendar.MONTH) + 1;
        this.year = calendar.get(Calendar.YEAR);
        peg(calendar);
    }

    /**
     * Returns the year in which the month falls.
     *
     * @return The year in which the month falls (as a Year object).
     */
    public Year getYear() {
        return new Year(this.year);
    }

    /**
     * Returns the year in which the month falls.
     *
     * @return The year in which the month falls (as an int).
     */
    public int getYearValue() {
        return this.year;
    }

    /**
     * Returns the month.  Note that 1=JAN, 2=FEB, ...
     *
     * @return The month.
     */
    public int getMonth() {
        return this.month;
    }

    /**
     * Returns the first millisecond of the month.  This will be determined 
     * relative to the time zone specified in the constructor, or in the 
     * calendar instance passed in the most recent call to the 
     * {@link #peg(Calendar)} method.
     *
     * @return The first millisecond of the month.
     * 
     * @see #getLastMillisecond()
     */
    public long getFirstMillisecond() {
        return this.firstMillisecond;
    }

    /**
     * Returns the last millisecond of the month.  This will be 
     * determined relative to the time zone specified in the constructor, or
     * in the calendar instance passed in the most recent call to the 
     * {@link #peg(Calendar)} method.
     *
     * @return The last millisecond of the month.
     * 
     * @see #getFirstMillisecond()
     */
    public long getLastMillisecond() {
        return this.lastMillisecond;
    }
    
    /** 
     * Recalculates the start date/time and end date/time for this time period 
     * relative to the supplied calendar (which incorporates a time zone).
     * 
     * @param calendar  the calendar (<code>null</code> not permitted).
     * 
     * @since 1.0.3
     */
    public void peg(Calendar calendar) {
        this.firstMillisecond = getFirstMillisecond(calendar);
        this.lastMillisecond = getLastMillisecond(calendar);
    }

    /**
     * Returns the month preceding this one.
     *
     * @return The month preceding this one.
     */
    public RegularTimePeriod previous() {
        Month result;
        if (this.month != MonthConstants.JANUARY) {
            result = new Month(this.month - 1, this.year);
        }
        else {
            if (this.year > 1900) {
                result = new Month(MonthConstants.DECEMBER, this.year - 1);
            }
            else {
                result = null;
            }
        }
        return result;
    }

    /**
     * Returns the month following this one.
     *
     * @return The month following this one.
     */
    public RegularTimePeriod next() {
        Month result;
        if (this.month != MonthConstants.DECEMBER) {
            result = new Month(this.month + 1, this.year);
        }
        else {
            if (this.year < 9999) {
                result = new Month(MonthConstants.JANUARY, this.year + 1);
            }
            else {
                result = null;
            }
        }
        return result;
    }

    /**
     * Returns a serial index number for the month.
     *
     * @return The serial index number.
     */
    public long getSerialIndex() {
        return this.year * 12L + this.month;
    }

    /**
     * Returns a string representing the month (e.g. "January 2002").
     * <P>
     * To do: look at internationalisation.
     *
     * @return A string representing the month.
     */
    public String toString() {
        return SerialDate.monthCodeToString(this.month) + " " + this.year;
    }

    /**
     * Tests the equality of this Month object to an arbitrary object.
     * Returns true if the target is a Month instance representing the same
     * month as this object.  In all other cases, returns false.
     *
     * @param obj  the object (<code>null</code> permitted).
     *
     * @return <code>true</code> if month and year of this and object are the 
     *         same.
     */
    public boolean equals(Object obj) {

        if (obj != null) {
            if (obj instanceof Month) {
                Month target = (Month) obj;
                return (this.month == target.getMonth() 
                        && (this.year == target.getYearValue()));
            }
            else {
                return false;
            }
        }
        else {
            return false;
        }

    }

    /**
     * Returns a hash code for this object instance.  The approach described by
     * Joshua Bloch in "Effective Java" has been used here:
     * <p>
     * <code>http://developer.java.sun.com/developer/Books/effectivejava
     * /Chapter3.pdf</code>
     * 
     * @return A hash code.
     */
    public int hashCode() {
        int result = 17;
        result = 37 * result + this.month;
        result = 37 * result + this.year;
        return result;
    }

    /**
     * Returns an integer indicating the order of this Month object relative to
     * the specified
     * object: negative == before, zero == same, positive == after.
     *
     * @param o1  the object to compare.
     *
     * @return negative == before, zero == same, positive == after.
     */
    public int compareTo(Object o1) {

        int result;

        // CASE 1 : Comparing to another Month object
        // --------------------------------------------
        if (o1 instanceof Month) {
            Month m = (Month) o1;
            result = this.year - m.getYearValue();
            if (result == 0) {
                result = this.month - m.getMonth();
            }
        }

        // CASE 2 : Comparing to another TimePeriod object
        // -----------------------------------------------
        else if (o1 instanceof RegularTimePeriod) {
            // more difficult case - evaluate later...
            result = 0;
        }

        // CASE 3 : Comparing to a non-TimePeriod object
        // ---------------------------------------------
        else {
            // consider time periods to be ordered after general objects
            result = 1;
        }

        return result;

    }

    /**
     * Returns the first millisecond of the month, evaluated using the supplied
     * calendar (which determines the time zone).
     *
     * @param calendar  the calendar (<code>null</code> not permitted).
     *
     * @return The first millisecond of the month.
     *
     * @throws NullPointerException if <code>calendar</code> is 
     *     <code>null</code>.
     */
    public long getFirstMillisecond(Calendar calendar) {
        calendar.set(this.year, this.month - 1, 1, 0, 0, 0);
        calendar.set(Calendar.MILLISECOND, 0);
        // in the following line, we'd rather call calendar.getTimeInMillis()
        // to avoid object creation, but that isn't supported in Java 1.3.1
        return calendar.getTime().getTime();
    }

    /**
     * Returns the last millisecond of the month, evaluated using the supplied
     * calendar (which determines the time zone).
     *
     * @param calendar  the calendar (<code>null</code> not permitted).
     *
     * @return The last millisecond of the month.
     *
     * @throws NullPointerException if <code>calendar</code> is 
     *     <code>null</code>.
     */
    public long getLastMillisecond(Calendar calendar) {
        int eom = SerialDate.lastDayOfMonth(this.month, this.year);
        calendar.set(this.year, this.month - 1, eom, 23, 59, 59);
        calendar.set(Calendar.MILLISECOND, 999);
        // in the following line, we'd rather call calendar.getTimeInMillis()
        // to avoid object creation, but that isn't supported in Java 1.3.1
        return calendar.getTime().getTime();
    }

    /**
     * Parses the string argument as a month.
     * <P>
     * This method is required to accept the format "YYYY-MM".  It will also
     * accept "MM-YYYY". Anything else, at the moment, is a bonus.
     *
     * @param s  the string to parse.
     *
     * @return <code>null</code> if the string is not parseable, the month 
     *         otherwise.
     */
    public static Month parseMonth(String s) {

        Month result = null;
        if (s != null) {

            // trim whitespace from either end of the string
            s = s.trim();

            int i = Month.findSeparator(s);
            if (i != -1) {
                String s1 = s.substring(0, i).trim();
                String s2 = s.substring(i + 1, s.length()).trim();

                Year year = Month.evaluateAsYear(s1);
                int month;
                if (year != null) {
                    month = SerialDate.stringToMonthCode(s2);
                    if (month == -1) {
                        throw new TimePeriodFormatException(
                            "Can't evaluate the month."
                        );
                    }
                    result = new Month(month, year);
                }
                else {
                    year = Month.evaluateAsYear(s2);
                    if (year != null) {
                        month = SerialDate.stringToMonthCode(s1);
                        if (month == -1) {
                            throw new TimePeriodFormatException(
                                "Can't evaluate the month."
                            );
                        }
                        result = new Month(month, year);
                    }
                    else {
                        throw new TimePeriodFormatException(
                            "Can't evaluate the year."
                        );
                    }
                }

            }
            else {
                throw new TimePeriodFormatException(
                    "Could not find separator."
                );
            }

        }
        return result;

    }

    /**
     * Finds the first occurrence of ' ', '-', ',' or '.'
     *
     * @param s  the string to parse.
     * 
     * @return <code>-1</code> if none of the characters where found, the
     *      position of the first occurence otherwise.
     */
    private static int findSeparator(String s) {

        int result = s.indexOf('-');
        if (result == -1) {
            result = s.indexOf(',');
        }
        if (result == -1) {
            result = s.indexOf(' ');
        }
        if (result == -1) {
            result = s.indexOf('.');
        }
        return result;
    }

    /**
     * Creates a year from a string, or returns null (format exceptions 
     * suppressed).
     *
     * @param s  the string to parse.
     *
     * @return <code>null</code> if the string is not parseable, the year 
     *         otherwise.
     */
    private static Year evaluateAsYear(String s) {

        Year result = null;
        try {
            result = Year.parseYear(s);
        }
        catch (TimePeriodFormatException e) {
            // suppress
        }
        return result;

    }

}
